### PR TITLE
Add internal lint for detecting non-glob imports of `rustc_type_ir::inherent`

### PR DESCRIPTION
**Pull request by fmease**
https://github.com/rust-lang/rust/pull/127627#issuecomment-2225295951

r? compiler-errors

### CODE REVIEW

File: compiler/rustc_lint/src/internal.rs [hunk 1]
```diff
@@ -263,6 +264,49 @@ fn gen_args(segment: &PathSegment<'_>) -> String {
...
+            {
+                (segment.ident.span, item.ident.span, "*")
+            }
+            [.., segment]
+                if path.res.iter().flat_map(Res::opt_def_id).any(is_mod_inherent)
+                    && let rustc_hir::UseKind::Single = kind =>
+            {
+                let (lo, snippet) =
+                    match cx.tcx.sess.source_map().span_to_snippet(path.span).as_deref() {
+                        Ok("self") => (path.span, "*"),
```
Comment (fmease) [inline compiler/rustc_lint/src/internal.rs:295]: Super gross but I couldn't think of anything better w/o tweaking AST lowering (e.g., to use `mark_with_reason` in this case). This is for dealing with `use path::to::{self};` properly. In HIR, there's no way to tell whether `use path;` came from `use path::{to};` or `use path::to::{self};`. Span doesn't help etc.

File: compiler/rustc_lint/src/internal.rs [hunk 2]
```diff
@@ -263,6 +264,49 @@ fn gen_args(segment: &PathSegment<'_>) -> String {
...
+declare_tool_lint! {
+    /// The `non_glob_import_of_type_ir_inherent_item` lint detects
+    /// non-glob imports of module `rustc_type_ir::inherent`.
+    pub rustc::NON_GLOB_IMPORT_OF_TYPE_IR_INHERENT,
+    Allow,
+    "non-glob import of `rustc_type_ir::inherent`",
+    report_in_external_macro: true
+}
+
+declare_lint_pass!(TypeIr => [NON_GLOB_IMPORT_OF_TYPE_IR_INHERENT]);
```
Comment (fmease) [inline compiler/rustc_lint/src/internal.rs:276]: I'm inclined to consolidate the lint passes `TypeIr` and `TyTyKind` to reduce bloat. Just need to think of a good name :thinking:

File: compiler/rustc_type_ir/src/effects.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-use crate::inherent::{AdtDef, IntoKind, Ty};
+use crate::inherent::*;
```
Comment (fmease) [inline compiler/rustc_type_ir/src/effects.rs:1]: :tada:

File: compiler/rustc_lint/src/internal.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_lint/src/internal.rs:276]: "opinionated imports"?

File: compiler/rustc_type_ir/src/effects.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/effects.rs:1]: omfg

**Review by compiler-errors — APPROVED** [review]
This is great, thanks

**Comment by fmease** [comment]
>  Lint didn't get registered?

It's registered … in a stage2 compiler but apparently we also run `tests/ui-fulldeps/` with a stage1 compiler during a full build. Didn't realize we test `tests/ui-fulldeps/` … twice?

@bors try

**Pull request merged**